crates/matrix-sdk/src/encryption/tasks.rs

```diff
@@ -169,70 +168,196 @@ impl BackupDownloadTask {
+    failures_cache: FailuresCache<RoomKeyInfo>,
 
-                if !already_tried && !task_exists && backups.are_enabled().await {
-                    let task = spawn(Self::download(
-                        client,
-                        room_key_info.to_owned(),
-                        failures_cache.to_owned(),
-                    ));
+    /// Map from event ID to download task
+    active_tasks: BTreeMap<OwnedEventId, JoinHandle<()>>,
 
-                    task_queue.insert(room_key_info, task);
-                }
-            } else {
-                trace!("Client got dropped, shutting down the task");
-                break;
-            }
+    /// A list of megolm sessions that we have already downloaded, or are about
+    /// to download.
+    ///
+    /// The idea here is that once we've (successfully) downloaded a session
+    /// from the backup, there's not much point trying again even if we get
+    /// another UTD event that uses the same session.
+    ///
+    /// TODO: that's not quite right though. In theory another client could
+    ///   update the backup with an earlier ratchet state, giving us access
+    ///   to earlier messages in the session. In which case, maybe this
+    ///   should expire?
+    downloaded_sessions: HashSet<RoomKeyInfo>,
+}
+
+impl BackupDownloadTaskListenerState {
+    /// Prepare a new `BackupDownloadTaskListenerState`.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - A reference to the `Client`, which is used to fire off the
+    ///   backup download request.
+    pub fn new(client: WeakClient) -> Self {
+        Self {
+            client,
+            failures_cache: FailuresCache::with_settings(Duration::from_secs(60 * 60 * 24), 60),
+            active_tasks: Default::default(),
+            downloaded_sessions: Default::default(),
         }
     }
+
+    /// Check if we should set off a download for the given request.
+    ///
+    /// Checks if:
+    ///  * we already have the key,
+    ///  * we have already downloaded this session, or are about to do so, or
+    ///  * we've backed off from trying to download this session.
+    ///
+    /// If any of the above are true, returns `false`. Otherwise, returns
+    /// `true`.
+    pub async fn should_download(&self, download_request: &RoomKeyDownloadRequest) -> bool {
+        // We should have a Client, because the calling function holds a strong
+        // reference to it.
+        let client = self.client.get().expect("Client unexpectedly dropped");
+
+        // Check that the Client has an OlmMachine
+        let machine_guard = client.olm_machine().await;
+        let Some(machine) = machine_guard.as_ref() else {
+            return false;
+        };
+
+        // Check if the keys for this message have arrived in the meantime
+        if machine
+            .is_room_key_available(&download_request.room_id, &download_request.megolm_session_id)
+            .await
+            .unwrap()
```

Comment:
Can we get also get rid of this `unwrap()` while we're here? I think unwrapping to a default value would be fine here, not sure if the default should be `true` or `false`.